sdk/go/common/resource/plugin/plugin.go

```diff
@@ -554,14 +555,33 @@ func buildPluginArguments(opts pluginArgumentOptions) []string {
 	return args
 }
 
+func (p *plugin) healthCheck() bool {
+	if p.Conn == nil {
+		return false
+	}
+
+	// Check that the plugin looks alive by calling gRPC's Health Check service.
+	// Most plugins don't actually implement this service, which is OK as we treat
+	// an unimplemented status as OK.
+	health := grpc_health_v1.NewHealthClient(p.Conn)
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	req := &grpc_health_v1.HealthCheckRequest{}
+	_, err := health.Check(ctx, req)
+	if err != nil && status.Code(err) != codes.Unimplemented {
+		logging.V(9).Infof("healthCheck(): failed with: %v", err)
+		return false
+	}
+
+	return true
+}
+
```

Comment:
```suggestion
func (p *plugin) healthCheck() bool {
	if p.Conn == nil {
		return false
	}

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	healthy := make(chan bool, 1)
	go func() {
		health := grpc_health_v1.NewHealthClient(p.Conn)
		req := &grpc_health_v1.HealthCheckRequest{}

		resp, err := health.Check(ctx, req)
		if err != nil {
			if status.Code(err) == codes.Unimplemented {
				healthy <- true 
				return
			}

			logging.V(9).Infof("healthCheck(): failed with: %v", err)
			healthy <- false
			return
		}

		healthy <- resp.Status == grpc_health_v1.HealthCheckResponse_SERVING
	}()

	select {
	case result := <-healthy:
		return result
	case <-ctx.Done(): // hit deadline
		return false
	}
}

```

- More explicit about handling timeouts through select pattern 
- validates the result from the health check
- early return inside of the error handling, easier to add more error cases later

Reply:
Thanks for the feedback, @cynicaljoy!